### ISSUE TITLE
[SCM] Using subfolder with a repository together with the conanfile

### ISSUE THREAD

**Issue by Adnn** [conan-io/conan#3069]
To help us debug your issue please explain:

- [x] I've read the [CONTRIBUTING guide](https://raw.githubusercontent.com/conan-io/conan/develop/.github/CONTRIBUTING.md).
- [x] I've specified the Conan version, operating system version and any tool that can be relevant.
- [x] I've explained the steps to reproduce the error or the motivation/use case of the question/suggestion.

Working with Conan 1.4.4 (also tested with 1.4.3) on OSX, we expect to rely on the (experimental) [SCM attribute of `ConanFile`](https://docs.conan.io/en/latest/creating_packages/package_repo.html#capturing-the-remote-and-commit-from-git-scm-experimental) to implement CI of our repositories (#3003 for context)

Here is a minimal conanfile:
```
from conans import ConanFile, CMake, tools

class ScmtestConan(ConanFile):
    name = "scm-test"
    version = "0.0.0"
    description = "Try the SCM conanfile attribute"
    settings = "os", "compiler", "build_type", "arch"

    generators = "cmake", "cmake_paths"

    scm = {
        "type": "git",
        "subfolder": "cloned_repo",
        "url": "auto",
        "revision": "auto"
    }

    def build(self):
        cmake = CMake(self)
        cmake.configure(source_folder="cloned_repo")
        cmake.build()
        cmake.install()

    def package(self):
        # Done by the install step in build()
        pass

    def package_info(self):
        self.cpp_info.libs = tools.collect_libs(self)
```

Now, if we export and try to install the recipe with `conan export . com/develop && conan install scm-test/0.0.0@com/develop`, the build fails.

### subfolder problem

After inspecting the created directories under `~/.conan`, it turn out that, even though a `cloned_repo` is created in the `source` folder, the repo is actually cloned directly into the `source` folder, and `cloned_repo` remains empty.

### same computer optimization problem

Now, if we were to move the conanfile.py from the root of the repo to a `conan/` subfolder, as we intend to, nothing is cloned at all, because it seems (following the TIP) that conan wants to use the local folder when installing on the same computer that did the export, and it uses the folder containing the conanfile, not necessarily the folder at the root of the repo.

**Comment by lasote** [comment]
Thanks for reporting, I'll have a look.

**Comment by lasote** [comment]
Hi! I cannot reproduce the issue. I saved your `conanfile.py` in a folder, and inside the folder I cloned a recipe repo at "cloned_repo" subfolder and then I run `conan export . com/develop && conan install scm-test/0.0.0@com/develop`.

The repository is created in the subfolder at the local cache and the build starts normally. Could you paste here the traces of the entire process?

Thanks!

**Comment by Adnn** [comment]
@lasote Of course, do I need to run the command a special way / look into a specific log file? Or do you just need the output in the terminal?

**Comment by Adnn** [comment]
Here is what is displayed in the terminal:
(please remember that I am on OSX - 10.12, in case it is os specific)

### Export

> $ conan export . com/develop
> WARN: Conanfile doesn't have 'url'.
> It is recommended to add it as attribute
> WARN: Conanfile doesn't have 'license'.
> It is recommended to add it as attribute
> scm-test/0.0.0@com/develop: Repo origin deduced by 'auto': git@github.com:Adnn/issue_conan-3069.git
> scm-test/0.0.0@com/develop: Revision deduced by 'auto': 2ee94f8181e70af0e6e0cb8d42a4e6a360e8fa68
> scm-test/0.0.0@com/develop: A new conanfile.py version was exported
> scm-test/0.0.0@com/develop: Folder: /Users/adn/.conan/data/scm-test/0.0.0/com/develop/export
> 
> $ git rev-parse --short HEAD
> 2ee94f8

### Install

$ conan install scm-test/0.0.0@com/develop --build
scm-test/0.0.0@com/develop: Installing package
Requirements
    scm-test/0.0.0@com/develop from local cache
Packages
    scm-test/0.0.0@com/develop:7fe67dff831b24bc4a8b5db678a51f1be5e44e7c

scm-test/0.0.0@com/develop: WARN: Forced build from source
scm-test/0.0.0@com/develop: Building your package in /Users/adn/.conan/data/scm-test/0.0.0/com/develop/build/7fe67dff831b24bc4a8b5db678a51f1be5e44e7c
scm-test/0.0.0@com/develop: Configuring sources in /Users/adn/.conan/data/scm-test/0.0.0/com/develop/source
scm-test/0.0.0@com/develop: Getting sources from folder: /Users/adn/projects/linux_todel/issue_conan-3069
scm-test/0.0.0@com/develop: Copying sources to build folder
scm-test/0.0.0@com/develop: Generator cmake created conanbuildinfo.cmake
scm-test/0.0.0@com/develop: Generator cmake_paths created conan_paths.cmake
scm-test/0.0.0@com/develop: Calling build()
CMake Error: The source directory "/Users/adn/.conan/data/scm-test/0.0.0/com/develop/build/7fe67dff831b24bc4a8b5db678a51f1be5e44e7c/cloned_repo" does not appear to contain CMakeLists.txt.
Specify --help for usage, or press the help button on the CMake GUI.
scm-test/0.0.0@com/develop:
scm-test/0.0.0@com/develop: ERROR: Package '7fe67dff831b24bc4a8b5db678a51f1be5e44e7c' build failed
scm-test/0.0.0@com/develop: WARN: Build folder /Users/adn/.conan/data/scm-test/0.0.0/com/develop/build/7fe67dff831b24bc4a8b5db678a51f1be5e44e7c
ERROR: scm-test/0.0.0@com/develop: Error in build() method, line 18
	cmake.configure(source_folder="cloned_repo")
	ConanException: Error 256 while executing cd '/Users/adn/.conan/data/scm-test/0.0.0/com/develop/build/7fe67dff831b24bc4a8b5db678a51f1be5e44e7c' && cmake -G "Unix Makefiles" -DCMAKE_BUILD_TYPE="Release" -DCONAN_EXPORTED="1" -DCONAN_COMPILER="apple-clang" -DCONAN_COMPILER_VERSION="9.0" -DCONAN_CXX_FLAGS="-m64" -DCONAN_SHARED_LINKER_FLAGS="-m64" -DCONAN_C_FLAGS="-m64" -DCONAN_LIBCXX="libc++" -DCMAKE_INSTALL_PREFIX="/Users/adn/.conan/data/scm-test/0.0.0/com/develop/package/7fe67dff831b24bc4a8b5db678a51f1be5e44e7c" -Wno-dev '/Users/adn/.conan/data/scm-test/0.0.0/com/develop/build/7fe67dff831b24bc4a8b5db678a51f1be5e44e7c/cloned_repo'

So install fails complaining .../cloned_repo does not contain the CMakeLists.txt, and for a good reason:

$ ls -1a /Users/adn/.conan/data/scm-test/0.0.0/com/develop/build/7fe67dff831b24bc4a8b5db678a51f1be5e44e7c/cloned_repo/
.
..
$ ls -1a /Users/adn/.conan/data/scm-test/0.0.0/com/develop/build/7fe67dff831b24bc4a8b5db678a51f1be5e44e7c/
.
..
.git
.gitignore
CMakeLists.txt
build
cloned_repo
conan_paths.cmake
conanbuildinfo.cmake
main.cpp

**Comment by Adnn** [comment]
I also tested in a Ubuntu (16.04) VM, and the situation is exactly the same as in OSX.

May be of interest:
I just realized that, on both test environments, executing `conan export` creates a `cloned_repo` folder in the folder containing the exported `conanfile.py`. I did not expect that directory creation behaviour, is it intentional?

**Comment by lasote** [comment]
Now I understand what happen. Let me think and I will come back to you.

**Comment by Adnn** [comment]
@lasote You may already be further than that, but after more investigation, I found that .conan/data/scm-test/0.0.0/com/develop/scm_folder.txt seems to be the file used for the "local repository clone" optimization.
After deleting it, the repository is correctly cloned under `.conan/data/scm-test/0.0.0/com/develop/source/cloned_repo/` instead of under the parent directory (and it is cloned from the remote git repository, instead of the local repo used for the export).

Of course, removing this file also solves the other complication described in the first post (making it mandatory to store the conanfile.py at the root of the repository).

**Comment by lasote** [comment]
Yes, I noticed that already. it is failing how the optimization copies the sources. Thanks!

**Comment by Adnn** [comment]
Is it possible to disable this optimization in Conan 1.4.4?
We don't really gain from it in our situation, and disabling it would solve two issues at once.

**Comment by lasote** [comment]
1.4.4 is already released, and it is not planned to release a minor before the end of the month (when 1.5 will be released). But let's think about it.

**Comment by lasote** [comment]
Please, remember that the SCM feature is experimental

**Comment by Adnn** [comment]
@lasote Thank you, sorry I was not clear: we are using 1.4.4, I was wondering if there was already some way to disable it in the released application  ; )

And yes, we know it is experimental so we will happily work around for the time being (as it is very useful for CI, but this is well established in other issues)

**Comment by Adnn** [comment]
Hi @lasote, it is great to hear you were able to solve this issue !

Perhaps there is a missunderstanding of what the `scm.subfolder` attribute is supposed to do. From the [1.4 documentation](https://docs.conan.io/en/latest/reference/conanfile/attributes.html?highlight=scm#scm), I understand it as equivalent to:

```
def source(self):
    self.run("git clone {scm.url} {scm.subfolder} && cd {scm.subfolder} && git reset --hard {scm.revision}")
```

The default generated `conanfile.py` lets git clone into the default [humanish subfolder](https://git-scm.com/docs/git-clone#git-clone-ltdirectorygt) provided by git, so it is equivalent to having `scm.subfolder=scm-test`. 
I wanted to keep a subfolder behaviour (not cloning directly into source, because when `conan install` copies sources to the build folder, I don't want a mess into `build/`), but I just chose a static name for the subfolder, because it makes writing some of our internal scripts easier. (i.e., our scripts don't have to generate the humanish name to `cd` into it)

**Comment by lasote** [comment]
Yes, and the `subfolder` is exactly for that, but when you are fetching an external repository, if you have the conanfile together with your code and you are capturing the scm by using "auto", for me, it makes no sense at all.

**Comment by Adnn** [comment]
@lasote Thank you for following up with that

I may be missing your point. Perhaps the incomprehension is with this

> But why would you need to call "conan sources" if your sources are already there to copy all of them to a subfolder?

When locally developing the recipe, we could of course just `cp` the sources to wherever we want. Even though up to now (i.e. before introducing `ConanFile.scm`in our recipes)  we preferred to actually `conan source` to test our implementation of the `source()` method.

Anyway, the fact is that the same recipe is intended to also be used to install the package by its reference (by the CI machine, by non-developpers that want to build one projects, etc...).
In this case of `conan install repo/0.5.3@com/stable`, we prefer the cloning to take place in the subfolder [not to litter the Conan's `build/` folder], and since the recipe is the same, the subfolder is also present when we locally develop the recipe.

**Comment by lasote** [comment]
> we preferred to actually conan source, in order to test our implementation of the source() method itself.

Totally agree, but typically using SCM the source method won't do much. Anyway, you can specify a `--source-folder` to another place and the source will be copied. But no subfolder is required for this.

> [not to litter the Conan's build/ folder root, but to isolate everything behind the subfolder]

What is everything? the same conanfile that will be in the subfolder too? or are you exporting some files? 

The issue should be solved at #3086 but still, I have some doubts about your approach makes sense. But anyway! if it works for you, it is ok!

**Comment by Adnn** [comment]
@lasote After some time... Thank you for fixing this issue!
